Add mines

WilliamQiufeng opened the pull request:
Requires https://github.com/Quaver/Quaver.API/pull/200
+ [ ] Requires a resource change

This PR adds visual and judgement code to mine notes introduced in the relevant Quaver.API PR. In the video below, mine notes are shown as red hollow rectangle notes.
When you happen to be pressing the lane between the marvelous window of the mine you will get a miss, otherwise a marvelous hit is generated

https://github.com/user-attachments/assets/4b5d1c34-fd71-4bae-830d-bd3cef1dd405

hehe

WilliamQiufeng committed:
add mines

ESV-Sweetplum commented:
I'm fine with this so long as all hitobjects are checked within range of receptors before the mines (as in prioritize a worse judgement over hitting a mine)

WilliamQiufeng commented:
> I'm fine with this so long as all hitobjects are checked within range of receptors before the mines (as in prioritize a worse judgement over hitting a mine)

Wait could you give an example? If I get it right wouldn't that mean that it could make the judgement window of the mine extremely small, or even zero, by putting a note right next to it?

ESV-Sweetplum commented:
Suppose at state.SongTime 200ms, there is a mine at 195ms and a note at 250ms, all in the same column. If an input is pressed, the note should take priority, even though the mine is closer to the receptor. This means that the player will receive a great judgement, and the mine will go by unnoticed.

ESV-Sweetplum commented:
> > I'm fine with this so long as all hitobjects are checked within range of receptors before the mines (as in prioritize a worse judgement over hitting a mine)
> 
> Wait could you give an example? If I get it right wouldn't that mean that it could make the judgement window of the mine extremely small, or even zero, by putting a note right next to it?

I suppose that's true, but personally I think it's on the mapper to avoid spamming mines such that this happens often

7ef38c commented:
> Suppose at state.SongTime 200ms, there is a mine at 195ms and a note at 250ms, all in the same column. If an input is pressed, the note should take priority, even though the mine is closer to the receptor. This means that the player will receive a great judgement, and the mine will go by unnoticed.

This implementation would make minedodge trivial no? While sure it is a more niche gimmick and either way,I don't think a player should just be able to ignore a mine if it would be in the note's great window, regardless of gimmicks.

ESV-Sweetplum commented:
maybe but i can only imagine the frustration if there's a mine like just a little bit before a note, and if you hit slightly early instead of a great you just shitmiss

edit: i'm also interested if `HitObjectType` is an exposed global as I don't see implementation for it in the files changed but i might just be tripping

Storn42 commented:
> maybe but i can only imagine the frustration if there's a mine like just a little bit before a note, and if you hit slightly early instead of a great you just shitmiss
> 
> edit: i'm also interested if `HitObjectType` is an exposed global as I don't see implementation for it in the files changed but i might just be tripping

whats stopping people from making a map with 1ms holds, or stacked notes, or a constant 100x sv? I dont think making a gameplay element trivial just because its sometimes annoying is the best way to go. some people find SV and LN frustrating to play as well, should we make LN tails count as long as you hold them until the marv window? I'm not saying mines need to have a massive hit window, but changing gameplay to make mine behavior to basically not matter isn't the way to go.

WilliamQiufeng commented:
> edit: i'm also interested if `HitObjectType` is an exposed global as I don't see implementation for it in the files changed but i might just be tripping

HitObjectType is declared in the Quaver.API PR (https://github.com/Quaver/Quaver.API/blob/0e52858d364f5e8a53f1311ccdc679b8c31985fb/Quaver.API/Enums/HitObjectType.cs)

ESV-Sweetplum commented:
> > maybe but i can only imagine the frustration if there's a mine like just a little bit before a note, and if you hit slightly early instead of a great you just shitmiss
> > edit: i'm also interested if `HitObjectType` is an exposed global as I don't see implementation for it in the files changed but i might just be tripping
> 
> whats stopping people from making a map with 1ms holds, or stacked notes, or a constant 100x sv? I dont think making a gameplay element trivial just because its sometimes annoying is the best way to go. some people find SV and LN frustrating to play as well, should we make LN tails count as long as you hold them until the marv window? I'm not saying mines need to have a massive hit window, but changing gameplay to make mine behavior to basically not matter isn't the way to go.

that's true, i was under the assumption that mines would have an identical hit window to normal notes. if that's not the case then it should be fine

WilliamQiufeng committed:
remove marv hit burst for mines
remove a log for debugging

WilliamQiufeng committed:
make ScoreProcessor.CalculateScore more consistent with its parameters

WilliamQiufeng committed:
Require isLnRelease and isMine parameter to be explicitly given in CalculateScore();

WilliamQiufeng committed:
Fix method ordering to fit develop branch

WilliamQiufeng committed:
Merge branch 'develop' into mines

# Conflicts:
#	Quaver.Shared/Screens/Gameplay/UI/Scoreboard/ScoreboardUser.cs
#	Quaver.Shared/Skinning/SkinKeys.cs

WilliamQiufeng committed:
Fix mine texture loading

WilliamQiufeng committed:
use the mine sprite of the first lane if the texture on the specified lane doesn't exist

WilliamQiufeng committed:
Use good mine icon for editor composition tool
Co-authored-by: Janko

WilliamQiufeng committed:
Fix mine loading for arrow skins

menvae commented on Quaver.Shared/Database/Maps/Map.cs:
```diff
@@ -186,14 +187,19 @@ public class Map
         /// </summary>
         public int LongNoteCount { get; set; }
 
+        /// <summary>
+        ///     The count of long notes.
+        /// </summary>
+        public int MineCount { get; set; }
```
typo in docs

WilliamQiufeng committed:
Merge remote-tracking branch 'origin/develop' into mines

WilliamQiufeng committed:
Fix hold hitobject not loaded correctly due to unwanted fillWithFirstIfNotExist=true in LoadHitObjects()

WilliamQiufeng committed:
For regular, hold start, body and end position Y and LongNoteSizeAdjustment, calculate such value for each note type.
Update GameplayHitObjectKeys so that the value of the right type is used when a hitobject is linked.

WilliamQiufeng committed:
Fix typo

AiAe commented:
@codex review

WilliamQiufeng committed:
Use mine specific fallbacks instead of regular notes'

WilliamQiufeng committed:
Register HitObjectType in Lua as hitobject_type

WilliamQiufeng committed:
Allow placing LN mines by mouse

WilliamQiufeng committed:
Don't make the normal hit sound if a mine is hit

Warp9000 committed:
update submodules
